scenarios/smart_trailer/applications/smart_trailer_application/src/main.rs

```diff
@@ -0,0 +1,252 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+// SPDX-License-Identifier: MIT
+
+use std::env;
+
+use digital_twin_model::trailer_v1;
+use digital_twin_providers_common::constants::chariott::{
+    INVEHICLE_DIGITAL_TWIN_SERVICE_COMMUNICATION_KIND,
+    INVEHICLE_DIGITAL_TWIN_SERVICE_COMMUNICATION_REFERENCE, INVEHICLE_DIGITAL_TWIN_SERVICE_NAME,
+    INVEHICLE_DIGITAL_TWIN_SERVICE_NAMESPACE, INVEHICLE_DIGITAL_TWIN_SERVICE_VERSION,
+};
+use digital_twin_providers_common::constants::{
+    constraint_type, digital_twin_operation, digital_twin_protocol,
+};
+use digital_twin_providers_common::utils::{
+    discover_digital_twin_provider_using_ibeji, discover_service_using_chariott,
+};
+use env_logger::{Builder, Target};
+use invehicle_stack_interfaces::module::managed_subscribe::v1::managed_subscribe_client::ManagedSubscribeClient;
+use invehicle_stack_interfaces::module::managed_subscribe::v1::{
+    Constraint, SubscriptionInfoRequest, SubscriptionInfoResponse,
+};
+use log::{debug, info, LevelFilter};
+use paho_mqtt as mqtt;
+use tokio::signal;
+use tokio::task::JoinHandle;
+use tokio::time::{sleep, Duration};
+use tonic::{Request, Status};
+use uuid::Uuid;
+
+const FREQUENCY_MS_FLAG: &str = "freq_ms=";
+const MQTT_CLIENT_ID: &str = "smart-trailer-consumer";
+
+// TODO: These could be added in configuration
+const CHARIOTT_SERVICE_DISCOVERY_URI: &str = "http://0.0.0.0:50000";
+
+const DEFAULT_FREQUENCY_MS: u64 = 10000; // 10 seconds
+
+// Constants used for retry logic
+const MAX_RETRIES: i32 = 10; // for demo purposes we will retry a maximum of 10 times
+                             // By default we will wait 5 seconds between retry attempts
+const DURATION_BETWEEN_ATTEMPTS: Duration = Duration::from_secs(5);
+
+/// Get trailer weight's subscription information from managed subscribe endpoint.
+///
+/// # Arguments
+/// * `managed_subscribe_uri` - The managed subscribe URI.
+/// * `constraints` - Constraints for the managed topic.
+async fn get_trailer_weight_subscription_info(
+    managed_subscribe_uri: &str,
+    constraints: Vec<Constraint>,
+) -> Result<SubscriptionInfoResponse, Status> {
+    // Create gRPC client.
+    let mut client = ManagedSubscribeClient::connect(managed_subscribe_uri.to_string())
+        .await
+        .map_err(|err| Status::from_error(err.into()))?;
+
+    let request = Request::new(SubscriptionInfoRequest {
+        entity_id: trailer_v1::trailer::trailer_weight::ID.to_string(),
+        constraints,
+    });
+
+    let response = client.get_subscription_info(request).await?;
+
+    Ok(response.into_inner())
+}
+
+/// Receive Trailer Weight updates.
+///
+/// # Arguments
+/// * `broker_uri` - The broker URI.
+/// * `topic` - The topic.
+async fn receive_trailer_weight_updates(
+    broker_uri: &str,
+    topic: &str,
+) -> Result<JoinHandle<()>, String> {
```

Comment:
To make use of Will's suggestion of using `?` update return type to `Result<JoinHandle<Result<(), String>>, String>`